faculty/clients/template.py

```diff
@@ -42,4 +45,310 @@ def publish_new(self, project_id, template, source_directory):
+    ):
+        endpoint = "project/{}/apply".format(target_project_id)
+        payload = {
+            "sourceProjectId": str(source_project_id),
+            "sourceDirectory": source_directory,
+            "targetDirectory": target_directory,
+            "parameterValues": parameters,
+        }
+        response = self._post_raw(endpoint, json=payload, check_status=False)
+        return _handle_publishing_response(response)
+
+
+def _handle_publishing_response(response):
+    if 200 <= response.status_code < 300:
+        return response
+    elif 400 <= response.status_code < 500:
+        response_body = response.json()
+        error_code = response_body.get("errorCode")
+        if error_code == "resources_validation_failure":
+            raise ResourceValidationFailureResponseSchema().load(response_body)
+        elif error_code == "parameter_validation_failure":
+            raise ParameterValidationFailureSchema().load(response_body)
+        elif error_code == "template_retrieval_failure":
+            raise TemplateRetrievalFailureResponseSchema().load(response_body)
+        elif error_code == "default_parameters_parsing_error":
+            raise DefaultParametersParsingErrorSchema().load(response_body)
+        elif error_code == "generic_parsing_failure":
+            raise GenericParsingErrorSchema().load(response_body)
+        elif error_code == "workspace_files_validation_error":
+            raise WorkspaceFilesValidationErrorSchema().load(response_body)
+    else:
+        raise Exception(
+            "Unexpected response from the server:\n", response.text
+        )
+
+
+class TemplateException(Exception):
+    pass
+
+
+FileTooLargeError = namedtuple(
+    "FileTooLargeError", ["path", "actual_size_bytes", "max_bytes"]
+)
+
+
+class FileTooLargeSchema(BaseSchema):
+    path = fields.String(required=True)
+    actual_size_bytes = fields.Int(data_key="actualSizeBytes", required=True)
+    max_bytes = fields.Int(data_key="maxBytes", required=True)
+
+    @post_load
+    def make_error(self, data):
+        return FileTooLargeError(**data)
+
+
+TooManyFilesError = namedtuple(
+    "TooManyFilesError", ["actual_files", "max_files"]
+)
+
+
+class TooManyFilesSchema(BaseSchema):
+    actual_files = fields.Int(data_key="actualFiles", required=True)
+    max_files = fields.Int(data_key="maxFiles", required=True)
+
+    @post_load
+    def make_error(self, data):
+        return TooManyFilesError(**data)
+
+
+class WorkpaceFilesValidationError(TemplateException):
+    def __init__(self, files_too_large, too_many_files):
+        self.files_too_large = files_too_large
+        self.too_many_files = too_many_files
+
+
+class WorkspaceFilesValidationErrorSchema(BaseSchema):
+    files_too_large = fields.List(
+        fields.Nested(FileTooLargeSchema()),
+        data_key="filesTooLarge",
+        required=True,
+    )
+    too_many_files = fields.Nested(
+        TooManyFilesSchema(), data_key="tooManyFiles"
+    )
```

Comment:
Is it correct that `filesTooLarge` is always set and that `tooManyFiles` is always optional? I don't know enough about the data model to know, but the discrepancy seems odd.

Reply:
That's correct. `tooManyFiles` is optional. `filesTooLarge` is a list, so the key is always present. No large files are represented as an empty list.